pkg/router/controller/route_secret_manager.go

```diff
@@ -118,106 +192,199 @@ func (p *RouteSecretManager) HandleRoute(eventType watch.EventType, route *route
 			secretNew := new.(*kapi.Secret)
-			log.V(4).Info("secret updated for route", "namespace", route.Namespace, "secret", secretNew.Name, "old-version", secretOld.ResourceVersion, "new-version", secretNew.ResourceVersion, "route", route.Name)
-
-			// re-validate
-			fldPath := field.NewPath("spec").Child("tls").Child("externalCertificate")
-			if err := routeapihelpers.ValidateTLSExternalCertificate(route, fldPath, p.sarClient, p.secretsGetter).ToAggregate(); err != nil {
-				log.Error(err, "skipping route due to invalid externalCertificate configuration", "namespace", route.Namespace, "route", route.Name)
-				p.recorder.RecordRouteRejection(route, "ExternalCertificateValidationFailed", err.Error())
-				p.plugin.HandleRoute(watch.Deleted, route)
-				return
-			}
+			key := generateKey(namespace, routeName)
+			msg := fmt.Sprintf("secret %q updated for route %q (old resourceVersion=%v, new resourceVersion=%v)", secretNew.Name, key, secretOld.ResourceVersion, secretNew.ResourceVersion)
+			log.V(4).Info(msg)
 
-			// read referenced secret (updated data)
-			secret, err := p.secretManager.GetSecret(context.TODO(), route.Namespace, route.Name)
+			// Ensure fetching the updated route
+			route, err := p.getUpdatedRoute(namespace, routeName)
 			if err != nil {
-				log.Error(err, "failed to get referenced secret")
-				p.recorder.RecordRouteRejection(route, "ExternalCertificateGetFailed", err.Error())
-				p.plugin.HandleRoute(watch.Deleted, route)
 				return
 			}
 
-			// update tls.Certificate and tls.Key
-			route.Spec.TLS.Certificate = string(secret.Data["tls.crt"])
-			route.Spec.TLS.Key = string(secret.Data["tls.key"])
-
-			// call the next plugin with watch.Modified
-			p.plugin.HandleRoute(watch.Modified, route)
-			// commit the changes
-			p.plugin.Commit()
+			// Update the route status, keeping `Admitted=True`, to signal route update
+			// and re-evaluatation by all the plugins (including this plugin).
+			p.recorder.RecordRouteUpdate(route, "ExternalCertificateSecretUpdated", msg)
 		},
 
 		DeleteFunc: func(obj interface{}) {
 			secret := obj.(*kapi.Secret)
-			msg := fmt.Sprintf("secret %s deleted for route %s/%s", secret.Name, route.Namespace, route.Name)
+			key := generateKey(namespace, routeName)
+			msg := fmt.Sprintf("secret %q deleted for route %q", secret.Name, key)
 			log.V(4).Info(msg)
 
-			// unregister associated secret monitor
-			if err := p.secretManager.UnregisterRoute(route.Namespace, route.Name); err != nil {
-				log.Error(err, "failed to unregister route")
+			// keep the secret monitor active and mark the secret as deleted for this route.
+			p.deletedSecrets.Store(key, true)
+
+			// Ensure fetching the updated route
+			route, err := p.getUpdatedRoute(namespace, routeName)
+			if err != nil {
+				return
 			}
 
+			// Reject this route
 			p.recorder.RecordRouteRejection(route, "ExternalCertificateSecretDeleted", msg)
+			// Stop serving this route
 			p.plugin.HandleRoute(watch.Deleted, route)
```

Comment:
Does it mean that the route whose external certificate was removed will be removed from the HAProxy config?